Record focus in route entry to move a11y focus to the last focused item

hannah-hyj opened the pull request:
issue: #97747 
engine pr:https://github.com/flutter/engine/pull/47114

## Pre-launch Checklist

- [ ] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [ ] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [ ] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [ ] I signed the [CLA].
- [ ] I listed at least one issue that this PR fixes in the description above.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [ ] I added new tests to check the change I am making, or this PR is [test-exempt].
- [ ] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

<!-- Links -->
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

chunhtai reviewed (commented):
Discussed offline, we should consider using the didGainAccessibility signel in the iOS and Android embedding to record the last focused semantics id, and use that to send focus event

hannah-hyj committed:
Update navigator.dart

Update navigatort.dart

1

Update navigator_test.dart

Update feedback.dart

Update navigator.dart

Update navigator.dart

lint

1

Update navigator_test.dart

Update feedback.dart

Update navigator_test.dart

Update navigator.dart

lint

1

hannah-hyj committed:
Update navigator.dart

hangyujin commented on packages/flutter/lib/src/widgets/navigator.dart:
```diff
@@ -3018,6 +3025,23 @@ class _RouteEntry extends RouteTransitionRecord {
   void handleDidPopNext(Route<dynamic> poppedRoute) {
     route.didPopNext(poppedRoute);
     lastAnnouncedPoppedNextRoute = WeakReference<Route<dynamic>>(poppedRoute);
+    if(lastFocusNode != null)
```
i think handleDidPopNext should only be called once even if there are multiple navigators.

chunhtai replied:
what if the lastFocusNode is outside of the inner navigator and the inner navigator didPopNext is called? Would this cause a re-announcement of the focused node?

hangyujin commented on packages/flutter/lib/src/widgets/navigator.dart:
```diff
@@ -3561,9 +3585,30 @@ class NavigatorState extends State<Navigator> with TickerProviderStateMixin, Res
       SystemNavigator.selectSingleEntryHistory();
     }
 
+    SystemChannels.accessibility.setMessageHandler((dynamic message) => handleAccessibilityMessage(message as Object));
```
just want to make sure i understand it right: 

1. setMessageHandler in `ServicesBinding`
2.  add something like `didReceiveAccessibilityMessage` to WidgetsBindingObserver
3.  call observer.didReceiveAccessibilityMessage() in `WidgetsBinding`
5. implement  WidgetsBindingObserver in Navigator

chunhtai replied:
Sorry I meant just general addListener/removelistener mechanism. WidgetBinderObserver may be too public facing for this api. I think you can create a similar structure as RelayoutWhenSystemFontsChangeMixin

hangyujin replied:
I still didn't get it.  RelayoutWhenSystemFontsChangeMixin is a mixin on RenderObject. By `create a similar structure as RelayoutWhenSystemFontsChangeMixin` do you mean creating a new mixin on ServicesBinding or a new mixin on Navigator?

Since ServicesBinding is already a mixin on BindingBase.  The closest thing I can think of is 
1. In Navigator, call  `ServicesBinding.instance.setAccessibilityMessageHandler()`
2.In ServicesBinding, call  `SystemChannels.accessibility.setMessageHandler()` in `setAccessibilityMessageHandler()`

hannah-hyj committed:
1

hannah-hyj committed:
resolve comments

hannah-hyj committed:
update test

hannah-hyj committed:
Update navigator.dart

hannah-hyj committed:
lint

chunhtai reviewed (approved):
LGTM

hannah-hyj committed:
Update navigator_test.dart

hannah-hyj committed:
Update navigator_test.dart

hannah-hyj committed:
update

hannah-hyj committed:
1
